Summary table updates

rabstejnek opened the pull request:
Made several updates to summary tables.

- Moved the "Save/Update" and "Cancel" buttons out of the "Overall" tab on the summary table forms, allowing them to be used regardless of which form tab is active.
- Added http://public.tableau.com to `valid_netloc_endings`.
- Updated JS method `urlify`:
  - It now takes `maxLength` and `suffixLength` parameters.
  - If the slugified version of the passed in string is > `maxLength`, then it is truncated and a random string is appended to the slug.  This random suffix is to make it more likely that the slug will be unique in the case that the unique portion of the slug was truncated off.

Example urlify:
```javascript
window.app.HAWCUtils.urlify('this is a short string')
'this-is-a-short-string'

window.app.HAWCUtils.urlify('this is a long string this is a long string this is a long string')
'this-is-a-long-string-this-is-a-long-string-t-ac3f'
'this-is-a-long-string-this-is-a-long-string-t-6206'
'this-is-a-long-string-this-is-a-long-string-t-9edd'
```

_Other changes:_
- Made `FormActions` for reusable "Submit"/"Cancel" buttons and implemented them where appropriate.
- Fixed rendering of invalid exploratory heatmaps, and added Actions so that users can fix them.
  - To reproduce test case, create a new exploratory heatmap without selecting a dataset.

shapiromatron commented on frontend/summary/summaryTable/editing/Root.js:
```diff
@@ -10,27 +10,38 @@ import DjangoForm from "./DjangoForm";
+                        <Table store={tableStore} forceReadOnly={true} />
+                    </TabPanel>
+                </Tabs>
+                <div className="form-actions">
```
can we make this div a reusable react component?  Maybe we can call it a `FormAction`?

It looks like there are currently 3 that already exist `BaseFormVisualReact, App, and DjangoForm`

The default for saveBtnText can be "Save"

rabstejnek replied:
I found another case, `frontend/mgmt/TaskTable/containers/Root.js`, where I was able to replace the form actions. I also replaced all the other instances, except for `BaseFormVisualReact`; this belongs to the aggregate endpoints visual form, which I was not able to test due to existing errors.

shapiromatron commented on hawc/apps/assessment/templates/assessment/effecttag_form.html:
```diff
@@ -16,7 +16,7 @@
     $(document).ready(function(){
       $('#id_name').on('keyup', function(){
-        $('#id_slug').val(window.app.HAWCUtils.urlify($(this).val(), 50));
+        $('#id_slug').val(window.app.HAWCUtils.urlify($(this).val(),128));
```
should the default length be 50? https://docs.djangoproject.com/en/3.2/ref/models/fields/#slugfield

shapiromatron replied:
doh; i think you did do that, and i'm looking at a different case. my bad.

shapiromatron commented on frontend/shared/utils/HAWCUtils.js:
```diff
@@ -203,8 +203,22 @@ class HAWCUtils {
 
-    static urlify(str) {
-        return slugify(str, {remove: /[^\w\s-_]/g});
+    static randomString(length) {
```
very similar code already exists here w/ a different implementation https://github.com/shapiromatron/hawc/blob/e54a8bee877e2cd63d03109106c3c108be3bf197/frontend/shared/utils/helpers.js#L246-L250

rabstejnek replied:
A quick benchmark shows the newer implementation being a bit faster (I imagine since its not relying on a regex):
![image](https://user-images.githubusercontent.com/46504563/133303865-2025138e-e135-4d06-9a77-61333f5651ce.png)
Since it'll also be helpful if this function can return a string variable in length, I'm going to replace this old method with the new one, and import it in to `HAWCUtils` for use.

shapiromatron commented on frontend/shared/utils/HAWCUtils.js:
```diff
@@ -203,8 +203,22 @@ class HAWCUtils {
+            .join("");
+    }
+
+    static urlify(str, maxLength = 50, suffixLength = 4) {
```
cool, i forget about how default params are allowed now...

shapiromatron reviewed (changes requested):
Looking good but a few more changes

rabstejnek commented:
@shapiromatron I sneaked a few other small fixes in this PR that I came across during testing. I've documented them in my initial comment.

rabstejnek commented:
@shapiromatron All of the requested changes are done, except for adding the `FormActions` component to `BaseFormVisualReact`; this belongs to the endpoint aggregation form, which I was not able to test because there are existing errors on submit. After these errors are addressed, it should be as easy as adding `<FormActions isForm={true} submitText="Save" cancelURL={this.config.cancel_url} />`. This is the one instance where the form actions actually take place within a form, which is why the `isForm` prop is present, and why we don't have to pass it a `handleSubmit`.

Two other things I've noticed:
- [This](https://github.com/shapiromatron/hawc/blob/summary-table-updates/frontend/summary/summaryForms/BaseVisualFormReact.js#L125) takes a `ref`, but from what I can see one is never created. It may be unnecessary.
- [This](https://github.com/shapiromatron/hawc/blob/summary-table-updates/frontend/summary/summaryForms/BaseVisualForm.js) seems to be a non-React version of this form, and from what I can tell its not used anywhere. We may be able to remove this file entirely.

shapiromatron commented:
> @shapiromatron All of the requested changes are done, except for adding the `FormActions` component to `BaseFormVisualReact`; this belongs to the endpoint aggregation form, which I was not able to test because there are existing errors on submit. After these errors are addressed, it should be as easy as adding `<FormActions isForm={true} submitText="Save" cancelURL={this.config.cancel_url} />`. This is the one instance where the form actions actually take place within a form, which is why the `isForm` prop is present, and why we don't have to pass it a `handleSubmit`.
> 
> Two other things I've noticed:
> 
> * [This](https://github.com/shapiromatron/hawc/blob/summary-table-updates/frontend/summary/summaryForms/BaseVisualFormReact.js#L125) takes a `ref`, but from what I can see one is never created. It may be unnecessary.
> * [This](https://github.com/shapiromatron/hawc/blob/summary-table-updates/frontend/summary/summaryForms/BaseVisualForm.js) seems to be a non-React version of this form, and from what I can tell its not used anywhere. We may be able to remove this file entirely.

Working on this now. To your second point; it looks like `summaryForms/BaseVisualForm.js` is used in multiple places
<img width="473" alt="Screen Shot 2021-09-20 at 6 43 33 PM" src="https://user-images.githubusercontent.com/999952/134086170-55362dd9-6868-4d15-b94d-d1874b7a8bb4.png">
?

shapiromatron commented on frontend/shared/utils/helpers.js:
```diff
@@ -6,6 +6,7 @@ import _ from "lodash";
 import moment from "moment";
 
 const stopwords = new Set("the is at which of on".split(" ")),
+    hexChars = "abcdef0123456789",
```
switched to hex characters instead of all for readability

shapiromatron commented on frontend/shared/utils/HAWCUtils.js:
```diff
@@ -203,8 +204,14 @@ class HAWCUtils {
+        // with a random string (to ensure uniqueness if uniqueness is truncated)
+        var slug = slugify(str, {remove: /[^\w\s-_]/g});
+        return slug.length > maxLength
+            ? slug.slice(0, maxLength - suffixLength - 1) + "-" + h.randomString(suffixLength)
```
added a hyphen between the slug and the entropy

shapiromatron commented on frontend/summary/summary/ExploreHeatmap.js:
```diff
@@ -178,7 +177,7 @@ class ExploreHeatmap extends BaseVisual {
             $plotDiv = $("<div>"),
             callback = resp => {
-                if (resp.dataset) {
+                if (resp.dataset || resp.error) {
```
i changed how this was implemented; and just  carried w/ the standard path even if an error was logged, just because it looks like they were generally the same so we didnt have as much duplicate code

shapiromatron reviewed (approved):
Just like #506; made some minor changes. If you're happy with them, please feel free to merge!

rabstejnek commented:
> > @shapiromatron All of the requested changes are done, except for adding the `FormActions` component to `BaseFormVisualReact`; this belongs to the endpoint aggregation form, which I was not able to test because there are existing errors on submit. After these errors are addressed, it should be as easy as adding `<FormActions isForm={true} submitText="Save" cancelURL={this.config.cancel_url} />`. This is the one instance where the form actions actually take place within a form, which is why the `isForm` prop is present, and why we don't have to pass it a `handleSubmit`.
> > Two other things I've noticed:
> > 
> > * [This](https://github.com/shapiromatron/hawc/blob/summary-table-updates/frontend/summary/summaryForms/BaseVisualFormReact.js#L125) takes a `ref`, but from what I can see one is never created. It may be unnecessary.
> > * [This](https://github.com/shapiromatron/hawc/blob/summary-table-updates/frontend/summary/summaryForms/BaseVisualForm.js) seems to be a non-React version of this form, and from what I can tell its not used anywhere. We may be able to remove this file entirely.
> 
> Working on this now. To your second point; it looks like `summaryForms/BaseVisualForm.js` is used in multiple places?
> <img alt="Screen Shot 2021-09-20 at 6 43 33 PM" width="473" src="https://user-images.githubusercontent.com/999952/134086170-55362dd9-6868-4d15-b94d-d1874b7a8bb4.png">
> 
> Update - completed. I also fixed the `EndpointAggregationFormReact.js` and you're now able to update and create.
> 
> I think I found the file you meant and deleted it - `EndpointAggregationForm.js`. I also go the update form working again. Wow, this stuff is terribly complex. We should investigate the bets way to remove complexity here; either use the exploratory heatmap pattern, or switch over to API-based. I wonder if going API-based is better in the long run since we know better how to do it now that we've built `SummaryTableViewset` in the last year.

Sorry, I had meant `EndpointAggregationForm.js`!

rabstejnek commented:
Looks good!
